File: heap/heap.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#include "heap.h"
#include "./vector/vector.h"

Heap *initHeap()
{
  Heap *heap = (Heap *)malloc(sizeof(Heap));
  heap->array = initVector();
  heap->length = heap->array->length;

  return heap;
}

void insertHeap(Heap *h, int value)
{
  push(h->array, value);
  h->length++;
  siftUp(h, h->length - 1);
}

void siftUp(Heap *h, int index)
{
  int *heap = h->array->array;

  if (index < 0 && index >= h->length)
    return;

  if (heap[index] > heap[(index - 1) / 2])
  {
    int aux = heap[(index - 1) / 2];
    heap[(index - 1) / 2] = heap[index];
    heap[index] = aux;
    siftUp(h, (index - 1) / 2);
  }
}

int getMax(Heap *h)
{
  return h->array->array[0];
}

int getSize(Heap *h)
{
  return h->length;
}

int isEmptyHeap(Heap *h)
{
  if (h->length)
    return 1;
  else
    return 0;
}

int extractMax(Heap *h)
{
  int maxValue = h->array->array[0];

  h->array->array[0] = h->array->array[h->length - 1];
  pop(h->array);
  h->length--;

  siftDown(h, 0);

  return maxValue;
}

void siftDown(Heap *h, int index)
{
  int *heap = h->array->array;
  int maxIndex = index;

  if (index < 0 && index >= h->length)
    return;

  if ((2 * index) + 1 < h->length && heap[(2 * index) + 1] > heap[maxIndex])
    maxIndex = (2 * index) + 1;
  if ((2 * index) + 2 < h->length && heap[(2 * index) + 2] > heap[maxIndex])
    maxIndex = (2 * index) + 2;

  if (heap[index] < heap[maxIndex])
  {
    int aux = heap[maxIndex];
    heap[maxIndex] = heap[index];
    heap[index] = aux;
    siftDown(h, maxIndex);
  }
}

void removeFromHeap(Heap *h, int index)
{
  int *heap = h->array->array;

  heap[index] = INT_MAX;
  siftUp(h, index);
  extractMax(h);
}

void heapify(Heap *h)
{
  for (int i = (h->length - 1) / 2; i >= 0; i--)
    siftDown(h, i);
}

void heapSort(Vector *v)
{
  Heap *h = initHeap();
  Vector *aux = initVector();
  destroyVector(h->array);
  h->array = v;
  h->length = h->array->length;

  heapify(h);

  int len = h->length;
  for (int i = 0; i < len; i++)
    push(aux, extractMax(h));

  len = aux->length;
  push(v, *(aux->array));
  for (int i = 1; i < len; i++)
    prepend(v, *(aux->array + i));

  h->array = aux;
  destroyHeap(h);
}

void destroyHeap(Heap *h)
{
  destroyVector(h->array);
  free(h);
}

void printHeap(Heap *h)
{
  printVector(h->array);
}

File: heap/python/heap.py
import sys

class Heap():
  def __init__(self):
    self.heap_list = []
    self.length = 0
  
  def get_min(self):
    return self.heap_list[0]
  
  def get_size(self):
    return self.length
  
  def is_empty_heap(self):
    if(self.length):
      return False
    else:
      return True
    
  def sift_up(self, index: int):
    if index <= 0 or index >= self.length:
      return
    
    if self.heap_list[index] < self.heap_list[(index - 1) // 2]:
      aux = self.heap_list[(index - 1) // 2]
      self.heap_list[(index - 1) // 2] = self.heap_list[index]
      self.heap_list[index] = aux
      self.sift_up((index - 1) // 2)
  
  def insert_on_heap(self, value: int):
    self.heap_list.append(value)
    self.length += 1
    self.sift_up(self.length - 1)

  def sift_down(self, index: int):
    if index < 0 or index >= self.length:
      return
    
    min_index = index

    if (2*index) + 1 < self.length and self.heap_list[(2*index) + 1] < self.heap_list[min_index]:
       min_index = (2*index) + 1
    if (2*index) + 2 < self.length and self.heap_list[(2*index) + 2] < self.heap_list[min_index]:
       min_index = (2*index) + 2 

    if(self.heap_list[index] > self.heap_list[min_index]):
      aux = self.heap_list[min_index]
      self.heap_list[min_index] = self.heap_list[index]
      self.heap_list[index] = aux
      self.sift_down(min_index)
  
  def extract_min(self):
    max_value = self.heap_list[0]

    self.heap_list[0] = self.heap_list[self.length - 1]
    self.heap_list.pop()
    self.length -= 1

    self.sift_down(0)

    return max_value
    
  def remove_from_heap(self, index: int):
    self.heap_list[index] = -sys.maxsize - 1
    self.sift_up(index)
    self.extract_min()

  def heapify(self):
    for index in range((self.length - 1) // 2, -1, -1):
      self.sift_down(index)

  def print_heap(self):
    print(self.heap_list)
    
